File: eva-core/src/main/java/io/nerv/core/util/tree/TreeHelper.java
package io.nerv.core.util.tree;

import cn.hutool.core.util.StrUtil;
import io.nerv.core.mvc.entity.mybatis.BaseTreeEntity;

import java.util.ArrayList;
import java.util.List;

/**
 * 树形数据工具类
 * @author: S.PKAQ
 * @Datetime: 2018/5/10 21:58
 */
public class TreeHelper {
    /**
     * 将list转换为树结构
     * @param moduleEntitys
     * @return
     */
    public List<BaseTreeEntity> bulid(List<BaseTreeEntity> moduleEntitys) {

        List<BaseTreeEntity> trees = new ArrayList<>();

        for (BaseTreeEntity entity : moduleEntitys) {
            String pid = entity.getParentId();
            if (StrUtil.isBlank(pid) || "0".equals(pid)) {
                trees.add(entity);
            }

            for (BaseTreeEntity it : moduleEntitys) {
                if (entity.getId().equals(it.getParentId())) {
                    if (entity.getOriginChildren() == null) {
                        entity.setChildren(new ArrayList<>());
                    }
                    entity.getOriginChildren().add(it);
                }
            }
        }
        return trees;
    }


    /**
     * 组装路径
     * :注意: 这里如果子路径本身就是以与父路径相同开头的路径那么不会拼接父路径
     * @param parentPath
     * @param currentPath
     * @return
     */
    public static String assemblePath(String parentPath, String currentPath,String oldParentPath) {
        if(StrUtil.isNotBlank(oldParentPath)) {
            //把子路径里原来的父路径替换成现在的父路径
            currentPath=currentPath.replace(oldParentPath,parentPath);
        }else {
            parentPath = (null != parentPath && parentPath.startsWith("/")) ? parentPath : "/" + parentPath;
            if (null != currentPath && !currentPath.startsWith(parentPath)) {
                currentPath = currentPath.startsWith("/") ? currentPath : "/" + currentPath;
                return parentPath + currentPath;
            }
        }
        return currentPath;
    }
}

File: eva-core/src/main/kotlin/io/nerv/core/util/tree/TreeHelper.kt
package io.nerv.core.util.tree

import cn.hutool.core.util.StrUtil
import io.nerv.core.mvc.entity.mybatis.BaseTreeEntity
import java.util.*

/**
 * 树形数据工具类
 * @author: S.PKAQ
 * @Datetime: 2018/5/10 21:58
 */
class TreeHelper {
    /**
     * 将list转换为树结构
     * @param moduleEntitys
     * @return
     */
    fun bulid(moduleEntitys: List<BaseTreeEntity>): List<BaseTreeEntity> {
        val trees: MutableList<BaseTreeEntity> = ArrayList()
        for (entity in moduleEntitys) {
            val pid: String = entity.getParentId()
            if (StrUtil.isBlank(pid) || "0" == pid) {
                trees.add(entity)
            }
            for (it in moduleEntitys) {
                if (entity.getId().equals(it.getParentId())) {
                    if (entity.originChildren == null) {
                        entity.setChildren(ArrayList<E>())
                    }
                    entity.originChildren.add(it)
                }
            }
        }
        return trees
    }

    companion object {
        /**
         * 组装路径
         * :注意: 这里如果子路径本身就是以与父路径相同开头的路径那么不会拼接父路径
         * @param parentPath
         * @param currentPath
         * @return
         */
        @kotlin.jvm.JvmStatic
        fun assemblePath(parentPath: String?, currentPath: String?, oldParentPath: String?): String? {
            var parentPath = parentPath
            var currentPath = currentPath
            if (StrUtil.isNotBlank(oldParentPath)) {
                //把子路径里原来的父路径替换成现在的父路径
                currentPath = currentPath!!.replace(oldParentPath!!, parentPath!!)
            } else {
                parentPath = if (null != parentPath && parentPath.startsWith("/")) parentPath else "/$parentPath"
                if (null != currentPath && !currentPath.startsWith(parentPath)) {
                    currentPath = if (currentPath.startsWith("/")) currentPath else "/$currentPath"
                    return parentPath + currentPath
                }
            }
            return currentPath
        }
    }
}